File: features/system/system_hub_cog.py
# cogs/system/system_hub.py

# NOTE: Pycord migration: Pycord is a maintained fork of discord.py with the same API, but should be imported as 'import discord' and 'from discord.ext import commands'.
# For maintainers: If you need to use Pycord-specific features, refer to https://docs.pycord.dev/en/master/
import discord  # Pycord (discord.py compatible)
from discord.ext import commands
import asyncio
import os
import sys

from features.system.ui.dropdown import SystemHubPublicView
from shared.utils.headers import render_loading_embed, get_system_status_header


class SystemHub(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command(name="hub")
    async def hub(self, ctx: commands.Context):
        print(f"[DEBUG] !hub command called by {ctx.author} in {ctx.channel}")
        user = ctx.author

        # --- UNIVERSAL PANEL STYLE ---
        header = get_system_status_header(user).replace('```ansi', '').replace('```', '').strip()
        desc = (
            f"```ansi\n{header}\n[ SHADOW NEXUS MAINFRAME ]\n"
            "──────────────────────────────────────────────\n"
            "😈 Open System Hub    → Begin your journey\n"
            "🔁 Reopen System Hub  → Return to the nexus\n"
            "──────────────────────────────────────────────\n"
            '\"You place your hand upon the terminal.\"\n'
            "```"
        )
        embed = discord.Embed(
            description=desc,
            color=discord.Color.dark_teal()
        )
        embed.set_footer(text="Watcher Console Uplink • Stable")
        # Pass the bot instance to the view so it can be used for database operations
        view = SystemHubPublicView(bot=self.bot, user=None)
        try:
            message = await ctx.send(embed=embed, view=view)
            print(f"[DEBUG] Message sent successfully - Message ID: {message.id}")
        except Exception as e:
            print(f"[DEBUG] Error sending message: {e}")
        print(f"[DEBUG] !hub command completed")

    @commands.command(name="restart", aliases=["reboot"])
    @commands.is_owner()  # Only bot owner can use this command
    async def restart_bot(self, ctx: commands.Context):
        """Restart the bot (Owner only)"""
        print(f"[ADMIN] Restart command called by {ctx.author} ({ctx.author.id})")
        
        # Delete the command message for cleaner chat
        try:
            await ctx.message.delete()
        except (discord.NotFound, discord.Forbidden):
            pass
        
        # Send confirmation embed
        embed = discord.Embed(
            title="🔄 System Restart Initiated",
            description=(
                "```ansi\n"
                "\u001b[1;31m[ SHADOW NEXUS MAINFRAME ]\u001b[0m\n"
                "──────────────────────────────────────────────\n"
                "\u001b[1;33m⚠️  SYSTEM RESTART IN PROGRESS\u001b[0m\n"
                "\u001b[0;37m• Closing database connections...\u001b[0m\n"
                "\u001b[0;37m• Terminating Redis cache...\u001b[0m\n"
                "\u001b[0;37m• Shutting down bot processes...\u001b[0m\n"
                "\u001b[0;37m• Restarting application...\u001b[0m\n"
                "──────────────────────────────────────────────\n"
                "\u001b[1;32mBot will be back online shortly.\u001b[0m\n"
                "```"
            ),
            color=discord.Color.orange()
        )
        embed.set_footer(text="System Administrator • Restart Protocol")
        
        restart_message = await ctx.send(embed=embed)
        
        # Wait a moment for the message to send
        await asyncio.sleep(2)
        
        # Create restart signal file for the wrapper script
        import pathlib
        restart_signal = pathlib.Path("restart_signal")
        restart_signal.touch()
        
        print("[ADMIN] Initiating bot restart...")
        
        # Close database connections gracefully
        if hasattr(self.bot, 'db_pool') and self.bot.db_pool:
            try:
                await self.bot.db_pool.close()
                print("[ADMIN] Database pool closed")
            except Exception as e:
                print(f"[ADMIN] Error closing database pool: {e}")
        
        # Close Redis connections gracefully
        if hasattr(self.bot, 'redis') and self.bot.redis:
            try:
                await self.bot.redis.close()
                print("[ADMIN] Redis connection closed")
            except Exception as e:
                print(f"[ADMIN] Error closing Redis connection: {e}")
        
        # Close the bot
        await self.bot.close()
        
        # Exit the process - the wrapper will restart it
        print("[ADMIN] Exiting process for restart...")
        sys.exit(0)


async def setup(bot: commands.Bot):
    await bot.add_cog(SystemHub(bot))

File: src/features/system/systemHubCog.ts
import { Message, EmbedBuilder, ActionRowBuilder, StringSelectMenuBuilder, StringSelectMenuOptionBuilder } from 'discord.js';
import { RealmBot } from '../../main';
import { getSystemStatusHeader } from '../../shared/utils/headers';

/**
 * System Hub Cog - Main navigation hub for the bot
 */
export class SystemHubCog {
    private bot: RealmBot;

    constructor(bot: RealmBot) {
        this.bot = bot;
    }

    /**
     * Handle !hub command
     */
    async handleHubCommand(message: Message): Promise<void> {
        console.log(`[DEBUG] !hub command called by ${message.author.tag} in ${message.channel}`);
        const user = message.author;

        // Create universal panel style embed
        const header = getSystemStatusHeader(user).replace(/```ansi/g, '').replace(/```/g, '').trim();
        const description = 
            `\`\`\`ansi\n${header}\n[ SHADOW NEXUS MAINFRAME ]\n` +
            `──────────────────────────────────────────────\n` +
            `😈 Open System Hub    → Begin your journey\n` +
            `🔁 Reopen System Hub  → Return to the nexus\n` +
            `──────────────────────────────────────────────\n` +
            `"You place your hand upon the terminal."\n` +
            `\`\`\``;

        const embed = new EmbedBuilder()
            .setDescription(description)
            .setColor(0x2B5B5A) // dark_teal equivalent
            .setFooter({ text: 'Watcher Console Uplink • Stable' });

        // Create select menu for system hub
        const selectMenu = new StringSelectMenuBuilder()
            .setCustomId(`system_public:${user.id}`)
            .setPlaceholder('Initialize Shadow Nexus…')
            .addOptions(
                new StringSelectMenuOptionBuilder()
                    .setLabel('Open System Hub')
                    .setValue('open_hub')
                    .setEmoji('😈'),
                new StringSelectMenuOptionBuilder()
                    .setLabel('Reopen System Hub')
                    .setValue('reopen_hub')
                    .setEmoji('🔁')
            );

        const row = new ActionRowBuilder<StringSelectMenuBuilder>()
            .addComponents(selectMenu);

        try {
            if (message.channel.isSendable()) {
                const sentMessage = await message.channel.send({
                    embeds: [embed],
                    components: [row]
                });
                console.log(`[DEBUG] Message sent successfully - Message ID: ${sentMessage.id}`);
            }
        } catch (error) {
            console.error('[DEBUG] Error sending message:', error);
        }
        
        console.log('[DEBUG] !hub command completed');
    }

    /**
     * Register command handlers
     */
    registerHandlers(): void {
        this.bot.on('messageCreate', async (message) => {
            if (message.author.bot) return;
            if (message.content.toLowerCase() === '!hub') {
                await this.handleHubCommand(message);
            }
        });

        // Handle select menu interactions
        this.bot.on('interactionCreate', async (interaction) => {
            if (!interaction.isStringSelectMenu()) return;
            if (!interaction.customId.startsWith('system_public:')) return;

            const userId = interaction.customId.split(':')[1];
            if (userId && interaction.user.id !== userId) {
                await interaction.reply({ content: 'This is not your menu.', ephemeral: true });
                return;
            }

            const value = interaction.values[0];
            
            if (value === 'open_hub' || value === 'reopen_hub') {
                // Import and show the full system hub view
                const { renderHubEmbed } = await import('./ui/systemHubView');
                const hubEmbed = await renderHubEmbed(this.bot, interaction.user);
                
                // For now, send ephemeral response - full panel implementation would go here
                await interaction.reply({ 
                    embeds: [hubEmbed], 
                    ephemeral: true 
                });
            }
        });
    }
}

/**
 * Setup function to initialize the cog
 */
export async function setup(bot: RealmBot): Promise<void> {
    const cog = new SystemHubCog(bot);
    cog.registerHandlers();
    console.log('[OK] System Hub cog loaded');
}
